### PR TITLE
perf(nuxt,schema): normalise components + directories more efficiently

### PR DESCRIPTION
**Pull request by danielroe**
### 📚 Description

spotted while investing nuxt module hooks, this rewrites a lot of the logic in the components module to be more performant (preferring single loops rather than looping multiple times over the same array), ~~as well as to move the `components:dirs` hook into `modules:done`. This should be non-breaking and _much_ more performant but perhaps we can test this and revisit this as an experimental change if anyone on @nuxt/core has concerns.~~

I've added types to an `any` which highlighted a few edge cases we needed to handle, as well as some inaccurate types for the `components` option in the Nuxt config.

Additionally, I spotted we were sorting `components` (if an array) by length but not `components.dirs` (if an array). We're now treating both the same way.

### CODE REVIEW

**Comment by danielroe** [comment]
I'll move the change to when the hook is called into a separate PR so we can discuss properly.

**Comment by danielroe** [comment]
@coderabbitai generate docstrings

File: packages/nuxt/src/components/module.ts [hunk 1]
```diff
@@ -50,81 +59,68 @@ export default defineNuxtModule<ComponentsOptions>({
         : context.components
     }
 
+    // TODO: remove in Nuxt v5
     if (nuxt.options.experimental.normalizeComponentNames) {
```
Comment (huang-julien) [inline packages/nuxt/src/components/module.ts:63]: is it about removing the flag or the feature ?

Comment (danielroe) [reply packages/nuxt/src/components/module.ts:63]: it's about removing the flag - ie. hard-coding the behaviour

**Comment by blowsie** [comment]
I ran into an issue with nuxt 4.1.2 where shadcn Form (using Vee Validate) is broken.
https://www.shadcn-vue.com/docs/components/form.html

I cant be 100% sure this PR is the cause, but I think it maybe.

The issue is the `form/index.ts` does not have a default export and the `Form` component being provided by [nuxt vee validate](https://nuxt.com/modules/vee-validate) is no longer loaded

**Comment by danielroe** [comment]
@blowsie do you have a reproduction? 🙏

**Comment by blowsie** [comment]
> @blowsie do you have a reproduction? 🙏 

Not to hand, but if I get a minute I'll try and create one
